### PR TITLE
#13347: ObRoot: Use field if property does not exist

### LINKED ISSUE #13347: PSSessions with Containers throw NullReferenceException
Cmdlets that deal with PSSessions created on ContainerId throw a NullReferenceException when creating the session. This is true for `New-PSSession`, `Enter-PSSession` and `Invoke-Command`, among maybe others.

The stack trace is this:
````
Object reference not set to an instance of an object.
   at System.Management.Automation.Runspaces.ContainerProcess.GetContainerProperties()
   at Microsoft.PowerShell.Commands.PSExecutionCmdlet.CreateHelpersForSpecifiedContainerSession()
````

For me the line that fails is inside `GetContainerPropertiesInternal` of the class `System.Management.Automation.Runspaces.ContainerProcess` in `System.Management.Automation.dll`.

`ContainerObRoot = (string) computeSystemPropertiesType.GetProperty("ObRoot").GetValue(obj);`

## Steps to reproduce

To reproduce this a running container is required.

```powershell
New-PSSession -ContainerId <TheContainerId> -RunAsAdministrator
```

## Expected behavior

An instance of `PSSession` should be created.

## Actual behavior

A NullReferenceException is thrown.

## Environment data

```none
Name                           Value
----                           -----
PSVersion                      7.0.3
PSEdition                      Core
GitCommitId                    7.0.3
OS                             Microsoft Windows 10.0.18363
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by hemisphera**
# PR Summary

Use field `ObRoot` if property `ObRoot` does not exist in `GetContainerPropertiesInternal`.

## PR Context

Fix #13347

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs [hunk 1]
```diff
@@ -3443,7 +3443,15 @@ private void GetContainerPropertiesInternal()
                     //
                     if (RuntimeId == Guid.Empty)
                     {
-                        ContainerObRoot = (string)computeSystemPropertiesType.GetProperty("ObRoot").GetValue(computeSystemPropertiesHandle);
+                        var pi = computeSystemPropertiesType.GetProperty("ObRoot");
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs:3446]: Please look a comment above for RuntimeId and follow the same pattern - check the field first.

**Review by SteveL-MSFT — APPROVED** [review]
LGTM. Thanks!

@hemisphera, you validated this works manually with a private build?

File: src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs [hunk 2]
```diff
@@ -3443,7 +3443,24 @@ private void GetContainerPropertiesInternal()
...
+                            var obRootPropertyInfo = computeSystemPropertiesType.GetProperty("ObRoot");
+                            if (obRootPropertyInfo != null)
+                            {
+                                ContainerObRoot = obRootPropertyInfo.GetValue(computeSystemPropertiesHandle) as string;
+                            }
+                        }
+
+                        if (ContainerObRoot == null) 
+                        {
+                            throw new ArgumentNullException(nameof(ContainerObRoot));
```
Comment (PaulHigin) [inline src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs:3462]: I think we should throw the `RemotingErrorIdStrings.CannotGetHostInteropTypes` exception here, same as above.

File: src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs [hunk 3]
```diff
@@ -3443,7 +3443,24 @@ private void GetContainerPropertiesInternal()
                     //
                     if (RuntimeId == Guid.Empty)
                     {
-                        ContainerObRoot = (string)computeSystemPropertiesType.GetProperty("ObRoot").GetValue(computeSystemPropertiesHandle);
+                        var obRootfieldInfo = computeSystemPropertiesType.GetField("ObRoot");
```
Comment (PaulHigin) [inline src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs:3446]: Please add similar comment about Hyper-V team changing property to field, to make it clear to others maintaining this code.

**Review by PaulHigin — CHANGES_REQUESTED** [review]

**Review by PaulHigin — APPROVED** [review]
LGTM
Thanks!

**Comment by SteveL-MSFT** [comment]
The codefactor issues are non-issues, so I think this PR is ready to go

File: src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs [hunk 4]
```diff
@@ -3443,7 +3443,26 @@ private void GetContainerPropertiesInternal()
...
+                            ContainerObRoot = obRootFieldInfo.GetValue(computeSystemPropertiesHandle) as string;
+                        }
+                        else
+                        {
+                            var obRootPropertyInfo = computeSystemPropertiesType.GetProperty("ObRoot");
+                            if (obRootPropertyInfo != null)
+                            {
+                                ContainerObRoot = obRootPropertyInfo.GetValue(computeSystemPropertiesHandle) as string;
+                            }
+                        }
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs:3460]:
```suggestion
                        ContainerObRoot = (computeSystemPropertiesType.GetField("ObRoot")?.GetValue(computeSystemPropertiesHandle)
                            ?? computerSystemPropertiesType.GetProperty("ObRoot")?.Value(computeSystemPropertiesHandle)) as string;
```

**Comment by adityapatwardhan** [comment]
@hemisphera Please have a look at suggestion from @rjmholt

File: src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs (hunk 4 above)
Comment (hemisphera) [reply src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs:3460]: As much as I usually like to compress code like this, I have to disagree on this one. I intentionally used these `if`s and `var`s to keep my changes in line with the comments and the existing code above - for better readability.

**Review by rjmholt — APPROVED** [review]

**Review by adityapatwardhan — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@hemisphera Thank you for your contribution!

**Pull request merged**